clients/pkg/promtail/targets/objectstore/reader.go

```diff
@@ -0,0 +1,251 @@
+					r.sendLines(bytes, labels) // send any pending bytes
+					r.object.BytesRead += int64(len(bytes))
+					r.markPositionAndSize()
+					r.active.Store(false)
+					r.acknowledgeMessage()
+					return
+
+				}
+				level.Error(r.logger).Log("msg", "error in reading line", "object", r.object.Key)
+			}
+			r.sendLines(bytes, labels)
+		}
+	}
+
+}
+
+func (r *objectReader) acknowledgeMessage() {
+	if err := r.ackMessage(); err != nil {
+		level.Error(r.logger).Log("msg", "error in acknowldeging message", "err", err)
+	}
+}
+
+func (r *objectReader) sendLines(bytes []byte, labels model.LabelSet) {
+	if len(bytes) < 1 {
+		return
+	}
+	time := time.Now().UTC()
+	line := strings.TrimRight(string(bytes), "\r\n")
+
+	// add metrics
+	r.metrics.objectReadLines.WithLabelValues(r.object.Key, r.storeName).Inc()
+	r.metrics.objectLogLengthHistogram.WithLabelValues(r.object.Key, r.storeName).Observe(float64(len(bytes)))
+
+	r.handler.Chan() <- api.Entry{
+		Labels: labels,
+		Entry: logproto.Entry{
+			Timestamp: time,
+			Line:      line,
+		},
+	}
+	// update bytes read
+	r.object.BytesRead += int64(len(bytes))
+}
+func (r *objectReader) markPositionAndSize() {
+	r.readerMtx.Lock()
+	defer r.readerMtx.Unlock()
+
+	// update positions
+	positionKey := fmt.Sprintf("object-%s", r.object.Key)
+	positionValue := fmt.Sprintf("%s:%s", strconv.FormatInt(r.object.ModifiedAt.UnixNano(), 10), strconv.FormatInt(r.object.BytesRead, 10))
+	r.positions.PutString(positions.CursorKey(positionKey), positionValue)
+
+	r.metrics.objectReadBytes.WithLabelValues(r.object.Key, r.storeName).Set(float64(r.object.BytesRead))
+}
+
+func (r *objectReader) GetObject() (*bufio.Reader, error) {
+	reader, err := r.objectClient.GetObject(context.Background(), r.object.Key)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch object")
+	}
+
+	bufReader, err := r.getReader(reader)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get reader")
+	}
+
+	if r.object.BytesRead == 0 {
+		return bufReader, nil
+	}
+
+	// if the object is previously known, we have to skip the previously read bytes.
+	_, err = bufReader.Peek(int(r.object.BytesRead))
+	if err != nil {
+		level.Warn(r.logger).Log("msg", "object size is less than previous known size. object will be tailed from beginning", "object", r.object.Key)
+		r.positions.Remove(fmt.Sprintf("s3object-%s", r.object.Key))
+		r.object.BytesRead = 0
+		return bufReader, nil
+	}
+
+	// skip the bytes which we have already read
+	_, err = bufReader.Discard(int(r.object.BytesRead))
+	if err != nil {
+		return nil, err
+	}
+
+	level.Info(r.logger).Log("msg", "previously known object. object will be tailed from the previous known position", "object", r.object.Key, "position", r.object.BytesRead)
+	return bufReader, nil
+}
+
+func (r *objectReader) getReader(reader io.ReadCloser) (*bufio.Reader, error) {
+	// identify the file type
+	buf := [3]byte{}
```

Comment:
Usually files have attribute may be that's easier ? WDYT ?

Reply:
When we do `GetObject` we only get `io.ReadCloser` type, it's not saved to a file. So, we have to identify by reading from the buffer.